packages/core/src/domain/internalMonitoring.ts

```diff
@@ -96,27 +95,31 @@ export function resetInternalMonitoring() {
   monitoringConfiguration.batch = undefined
 }
 
-export function monitored(_: any, __: string, descriptor: PropertyDescriptor) {
-  const originalMethod = descriptor.value
-  descriptor.value = function () {
-    const decorated = (monitoringConfiguration.batch ? monitor(originalMethod) : originalMethod) as Function
-    return decorated.apply(this, arguments)
-  }
+export function monitored<T extends (...params: any[]) => unknown>(
+  _: any,
+  __: string,
+  descriptor: TypedPropertyDescriptor<T>
+) {
+  const originalMethod = descriptor.value!
+  descriptor.value = (function (this: any, ...args: Parameters<T>) {
+    const decorated = monitoringConfiguration.batch ? monitor(originalMethod) : originalMethod
+    return decorated.apply(this, args) as ReturnType<T>
+  } as any) as T
 }
 
-export function monitor<T extends Function>(fn: T): T {
-  return (function (this: any) {
+export function monitor<R>(fn: (...args: any[]) => R) {
```

Comment:
Monitored functions types are less accurate because of this. See `packages/rum/esm/boot/rum.entry.d.ts` (generated by `yarn build`):

```
export declare const datadogRum: {
    init: (...args: any[]) => void;
    addRumGlobalContext: (...args: any[]) => void;
    removeRumGlobalContext: (...args: any[]) => void;
    getRumGlobalContext: (...args: any[]) => import("@datadog/browser-core").Context;
    setRumGlobalContext: (...args: any[]) => void;
    getInternalContext: (...args: any[]) => import("@datadog/browser-rum-core/cjs/rawRumEvent.types").InternalContext | undefined;
    addAction: (...args: any[]) => void;
    addUserAction: (name: string, context?: object | undefined) => void;
    addError: (...args: any[]) => void;
    setUser: (...args: any[]) => void;
} & {
    onReady(callback: () => void): void;
};
```

What about restoring `T extends Function`? or maybe `T extends: (...args: any[]) => any` if `Function` was the issue